numbers.lisp

```diff
@@ -66,53 +66,52 @@ In C terms, this is a postdecrement while `decf' is a predecrement.")
 
 (defun read-float-aux (stream junk-allowed type)
   #.+merge-tail-calls+
-  (labels ((junk ()
-             (error "Junk in string"))
-           (next ()
-             (let ((char (read-char stream nil nil)))
-               (values char (and char (digit-char-p char)))))
-           (before-decimal (number)
-             (multiple-value-bind (char digit) (next)
-               (cond ((null char)
-                      (coerce number type))
-                     ((eql char #\.)
-                      (after-decimal number 1))
-                     (digit
-                      (before-decimal (+ (* number 10) digit)))
-                     ((exponent-char? char)
-                      (exponent number 1 0 (exponent-char-format char)))
-                     (t (if junk-allowed
-                            number
-                            (junk))))))
-           (after-decimal (number divisor)
-             (multiple-value-bind (char digit) (next)
-               (cond ((null char)
-                      (make-float number divisor type))
-                     ((exponent-char? char)
-                      (exponent number divisor 0
-                                (exponent-char-format char)))
-                     (digit
-                      (after-decimal (+ (* number 10) digit)
-                                     (* divisor 10)))
-                     (t (if junk-allowed
-                            (make-float number divisor)
-                            (junk))))))
-           (exponent (n d e f &optional neg)
-             (multiple-value-bind (char digit) (next)
-               (cond ((null char)
-                      (let ((e (if neg (- e) e)))
-                        (make-float (* (expt 10 e) n) d f)))
-                     ((eql char #\+)
-                      (exponent n d e f nil))
-                     ((eql char #\-)
-                      (exponent n d e f t))
-                     (digit
-                      (exponent n d (+ (* e 10) digit) f neg))
-                     (t (if junk-allowed
-                            (let ((e (if neg (- e) e)))
-                              (make-float (* (expt 10 e) n) d f))
-                            (junk)))))))
-    (before-decimal 0)))
+  (macrolet ((maybe-junk (&body body)
+               `(if junk-allowed
+                    (progn
+                      (unread-char char stream)
+                      ,@body)
+                    (error "Junk in string"))))
```

Comment:
I don't think a macro is necessarily the best way to do this. A simpler approach (and a smaller diff) might be to add something like a `junk-found` function; then `(if junk-allowed x)` becomes two forms, `(junk-found) x`.

Reply:
Thanks for the suggestion! I have revised the patch accordingly.